library/src/engine/engine.ts

```diff
@@ -14,19 +14,99 @@ import {
   type InitContext,
   type OnRemovalFn,
   PluginType,
-  type RemovalEntry,
   Requirement,
   type RuntimeContext,
   type RuntimeExpressionFunction,
   type WatcherPlugin,
 } from './types'
 
+const removalKey = (k: string, v: string) => `${k}${DSP}${v}`
+
 export class Engine {
   #signals = new SignalsRoot()
   #plugins: AttributePlugin[] = []
   #actions: ActionPlugins = {}
   #watchers: WatcherPlugin[] = []
-  #removals = new Map<Element, RemovalEntry>()
+
+  // Map of cleanup functions by element, keyed by the raw dataset key and value
+  #removals = new Map<Element, Map<string, OnRemovalFn>>()
+
+  constructor() {
+    const dsPrefix = 'data-'
+
+    const ob = new MutationObserver((mutations) => {
+      for (const {
+        target,
+        type,
+        attributeName,
+        oldValue,
+        addedNodes,
+        removedNodes,
+      } of mutations) {
+        const elify = (n: Node) => n as HTMLorSVGElement
```

Comment:
Don't know if you're interested in comments on this PR, nor do I know if you would find this helpful or not, but I would leave this comment here for consideration:

Perhaps this function could be moved outside of the for loop so it's not redeclared every iteration, as it doesn't depend on any context of the loop? 

Or, perhaps more extreme, since all `elify` is doing is a Typescript type assertion, is it worth just doing the assertion inline in the 4(?) places `elify` output is used, and thus avoid introducing any runtime code at all? Might be a bit uglier and a slightly poorer DX, but I think there's an argument to be made that it's kind of a pointless function call beyond making the Typescript compiler happy...

Reply:
valid point, fixed and it's a few bytes smaller!